src/coprocessor/statistics/fmsketch.rs

```diff
@@ -0,0 +1,164 @@
+            max_size: max_size,
+            hash_set: HashSet::new(),
+        }
+    }
+
+    // ndv returns the approximate number of distinct elements
+    pub fn ndv(&self) -> u64 {
+        (self.mask + 1) * (self.hash_set.len() as u64)
+    }
+
+    pub fn insert(&mut self, mut bytes: &[u8]) {
+        let hash = {
+            let mut out: [u8; 16] = [0; 16];
+            murmur3_x64_128(&mut bytes, 0, &mut out);
+            LittleEndian::read_u64(&out[0..8])
+        };
+        self.insert_hash_value(hash);
+    }
+
+    pub fn into_proto(self) -> analyze::FMSketch {
+        let mut proto = analyze::FMSketch::new();
+        proto.set_mask(self.mask);
+        let hash = self.hash_set.into_iter().collect();
+        proto.set_hashset(hash);
+        proto
+    }
+
+    fn insert_hash_value(&mut self, hash_val: u64) {
+        if (hash_val & self.mask) != 0 {
+            return;
+        }
+        self.hash_set.insert(hash_val);
+        if self.hash_set.len() > self.max_size {
+            let mask = (self.mask << 1) | 1;
+            self.hash_set.retain(|&x| x & mask == 0);
+            self.mask = mask;
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use coprocessor::codec::datum;
+    use coprocessor::codec::datum::Datum;
+    use coprocessor::codec::Result;
+    use util::as_slice;
+    use super::*;
+
+    struct TestData {
+        samples: Vec<Datum>,
+        rc: Vec<Datum>,
+        pk: Vec<Datum>,
+    }
+
+    fn generate_samples(count: usize) -> Vec<Datum> {
+        let mut samples = Vec::with_capacity(count);
+        samples.resize(count, 0);
+        let start = 1000;
+
+        for mut item in samples.iter_mut().take(start).skip(1) {
+            *item = 2;
+        }
+
+        for (id, mut item) in samples.iter_mut().enumerate().take(count).skip(start) {
+            *item = id;
+        }
+
+        let mut id = start;
+        while id < count {
+            samples[id] += 1;
+            id += 3;
+        }
+
+        id = start;
+        while id < count {
+            samples[id] += 2;
+            id += 5;
+        }
+        samples.into_iter().map(|v| Datum::I64(v as i64)).collect()
+    }
+
+    impl Default for TestData {
+        fn default() -> TestData {
+            let samples = generate_samples(10000);
+            let count = 100000;
+            let rc = generate_samples(count);
+
+            let mut pk = Vec::with_capacity(count);
+            pk.resize(count, Datum::Null);
+            for (id, mut item) in pk.iter_mut().enumerate().take(count) {
+                *item = Datum::I64(id as i64);
+            }
+            TestData {
+                samples: samples,
+                rc: rc,
+                pk: pk,
+            }
+        }
+    }
+
+    pub fn build_fmsketch(values: &[Datum], max_size: usize) -> Result<FMSketch> {
+        let mut s = FMSketch::new(max_size);
+        for value in values {
+            let bytes = try!(datum::encode_value(as_slice(value)));
```

Comment:
Why `as_slice`?

Reply:
Since `encode_value` needs a `&[Datum]` while value is `&Datum`

Reply:
Since `encode_value` needs a `&[Datum]` while value is `&Datum`  @BusyJay